### PR TITLE
add cupy.nonzero, cupy.flatnonzero, cupy.ix_

### PR DESCRIPTION
**Pull request by kikusu**
This PR is add some  cupy functions.
- cupy.nonzero
- cupy.flatnonzero
- cupy.ix_
- cupy.ndarray.nonzero

#### Impliment
1. src = [-1, 1, 0, 2]
2. candidate = src != 0 ==> \[ True, True, False, True \] (bool array) ==> \[1, 1, 0, 1\] (int array)
3. scan = inclusive_scan( candidate) ==> [ 1, 2, 2, 3 ]
4. for i in xrange(len(src) ): nonzero[ scan[n]-1 ] = n if src[n] != 0
5. nonzero ==> [ 0, 1, 3]

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -2303,3 +2337,172 @@ cdef _mean = create_reduction_func(
      ('e->e', (None, None, None, 'float')),
      'f->f', 'd->d'),
     ('in0', 'a + b', 'out0 = a / (_in_ind.size() / _out_ind.size())', None))
+
+# -----------------------------------------------------------------------------
+# scan
+# -----------------------------------------------------------------------------
+
+def _inclusive_scan_kernel(dtype, block_size):
```
Comment (okuta) [inline cupy/core/core.pyx]: How about using `cupy.memoize`.

Comment (kikusu) [inline cupy/core/core.pyx]: I did not know `cupy.memoize`.  I added it.

Comment (okuta) [inline cupy/core/core.pyx]: For example.
https://github.com/pfnet/chainer/blob/457550fc2501f3b8f73f93d69400f51d4096696f/cupy/core/reduction.pxi#L168

**Comment by kikusu** [comment]
update

**Comment by okuta** [comment]
Could you split modification of the coding style from this PR?
Please send it as another PR.

File: cupy/indexing/generate.py [hunk 2]
```diff
@@ -6,12 +6,64 @@
...
+    for k, new in enumerate(args):
+        new = cupy.asarray(new)
+        if new.ndim != 1:
+            raise ValueError("Cross index must be 1 dimensional")
+        if new.size == 0:
+            # Explicitly type empty arrays to avoid float default
+            new = new.astype(numpy.intp)
+        if cupy.issubdtype(new.dtype, cupy.bool_):
+             new, = new.nonzero()
+        new = new.reshape((1,)*k + (new.size,) + (1,)*(nd-k-1))
```
Comment (okuta) [inline cupy/indexing/generate.py]: Please add a space.

File: cupy/indexing/generate.py [hunk 3]
```diff
@@ -6,12 +6,64 @@
...
+def ix_(*args):
+    """Construct an open mesh from multiple sequences.
+
+    This function takes N 1-D sequences and returns N outputs with N
+    dimensions each, such that the shape is 1 in all but one dimension
+    and the dimension with the non-unit shape value cycles through all
+    N dimensions.
+
+    Using `ix_` one can quickly construct index arrays that will index
+    the cross product. ``a[np.ix_([1,3],[2,5])]`` returns the array
```
Comment (okuta) [inline cupy/indexing/generate.py]: Please fix `np` to `cupy`.

File: cupy/indexing/generate.py [hunk 4]
```diff
@@ -6,12 +6,64 @@
...
+    Args:
+        *args: 1-D sequences
+
+    Returns: tuple of ndarrays
+        N arrays with N dimensions each, with N the number of input
+        sequences. Together these arrays form an open mesh.
+
+    Examples
+    --------
+    >>> a = np.arange(10).reshape(2, 5)
```
Comment (okuta) [inline cupy/indexing/generate.py]: Please fix `np` to `cupy`.

File: cupy/indexing/generate.py [hunk 5]
```diff
@@ -6,12 +6,64 @@
...
+    >>> a
+    array([[0, 1, 2, 3, 4],
+           [5, 6, 7, 8, 9]])
+    >>> ixgrid = np.ix_([0,1], [2,4])
```
Comment (okuta) [inline cupy/indexing/generate.py]: Please fix `np` to `cupy`.

File: cupy/core/core.pyx [hunk 6]
```diff
@@ -2303,3 +2338,176 @@ cdef _mean = create_reduction_func(
...
+        }
+        """).substitute(name=name, src_dtype=src_dtype,
+                        src_ndim=src_ndim, index_dtype=index_dtype,
+                        dst_dtype=dst_dtype)
+    module = compile_with_cache(source)
+
+    return module.get_function(name)
+
+def scan(a):
+    if (a.ndim != 1):
```
Comment (okuta) [inline cupy/core/core.pyx]: Please remove parenthesize.

File: cupy/core/core.pyx [hunk 7]
```diff
@@ -2303,3 +2338,176 @@ cdef _mean = create_reduction_func(
...
+
+    source = string.Template("""
+        extern "C" __global__ void ${name}(const CArray<${src_dtype}, 1> src,
+            CIndexer<${src_ndim}> shape,
+            const CArray<${index_dtype}, 1> scaned_index,
+            CArray<${dst_dtype}, 1> dst){
+
+            int thid = blockIdx.x * blockDim.x + threadIdx.x;
+
+
```
Comment (okuta) [inline cupy/core/core.pyx]: Please remove this line.

**Comment by okuta** [comment]
Please add the test for `scan`.
Especially it is necessary to test the boundary condition and large input.

File: cupy/indexing/generate.py [hunk 8]
```diff
@@ -6,12 +6,64 @@
...
+    N dimensions.
+
+    Using `ix_` one can quickly construct index arrays that will index
+    the cross product. ``a[np.ix_([1,3],[2,5])]`` returns the array
+    ``[[a[1,2] a[1,5]], [a[3,2] a[3,5]]]``.
+
+    Args:
+        *args: 1-D sequences
+
+    Returns: tuple of ndarrays
```
Comment (okuta) [inline cupy/indexing/generate.py]: Please fix comment style.

**Comment by kikusu** [comment]
Thank you for review. Sorry for the many mistakes.I add scan test and remove modification of the coding.

**Comment by kikusu** [comment]
Sorry, I failed to reabse. I updated the commit.

File: cupy/core/core.pyx [hunk 9]
```diff
@@ -576,7 +576,43 @@ cdef class ndarray:
...
+        dtype = numpy.int64
+
+        scan_index = scan(condition.astype(dtype).ravel())
+        count_nonzero = int(scan_index[-1])
+
+        if self.ndim <= 1:
+            dst = ndarray((count_nonzero,), dtype=dtype)
+            kern = _nonzero_1d_kernel(self.dtype, dtype)
+
+            if self.ndim == 0:
```
Comment (okuta) [inline cupy/core/core.pyx]: How about removing this check?

**Comment by kikusu** [comment]
I remov 0D, 1D array check.

File: cupy/indexing/generate.py [hunk 10]
```diff
@@ -6,12 +6,65 @@
 # class r_(object):
 # class s_(object):
 
-
+import numpy
+import cupy
 # TODO(okuta): Implement indices
 
 
 # TODO(okuta): Implement ix_
```
Comment (okuta) [inline cupy/indexing/generate.py]: Please remove it.

File: tests/cupy_tests/core_tests/test_scan.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+from cupy import testing
+
+
+class TestScan(unittest.TestCase):
+
+    @testing.for_all_dtypes()
+    def test_scan(self, dtype):
+        element_num = 10000
+
+        if dtype in [cupy.int8, cupy.uint8]:
```
Comment (okuta) [inline tests/cupy_tests/core_tests/test_scan.py]: How about `{cupy.int8, cupy.uint8}`?

File: cupy/indexing/generate.py [hunk 12]
```diff
@@ -6,12 +6,65 @@
...
+    >>> a
+    array([[0, 1, 2, 3, 4],
+           [5, 6, 7, 8, 9]])
+    >>> ixgrid = cupy.ix_([0,1], [2,4])
+    >>> ixgrid
+    (array([[0],
+           [1]]), array([[2, 4]]))
+    >>> ixgrid[0].shape, ixgrid[1].shape
+    ((2, 1), (1, 2))
+    >>> a[ixgrid]
```
Comment (okuta) [inline cupy/indexing/generate.py]: This operation can not support in current CuPy.

File: tests/cupy_tests/sorting_tests/test_search.py [hunk 13]
```diff
@@ -123,3 +125,36 @@ def test_one_argument(self, xp):
...
+    {"array": numpy.array(1)},
+)
+@testing.gpu
+class TestNonzero(unittest.TestCase):
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_array_list_equal()
+    def test_nonzero(self, xp, dtype):
+        array = xp.array(self.array, dtype=dtype)
+
```
Comment (okuta) [inline tests/cupy_tests/sorting_tests/test_search.py]: Remove this line.

**Comment by okuta** [comment]
LGTM except comments.

**Comment by okuta** [comment]
OK, Thank you very much!

**Comment by kikusu** [comment]
Thank you for review for a long time. 🙇
